src/sinks/util/encoding.rs

```diff
@@ -0,0 +1,312 @@
+use std::io;
+
+use bytes::BytesMut;
+use codecs::encoding::Framer;
+use tokio_util::codec::Encoder as _;
+
+use crate::{codecs::Transformer, event::Event};
+
+pub trait Encoder<T> {
+    /// Encodes the input into the provided writer.
+    ///
+    /// # Errors
+    ///
+    /// If an I/O error is encountered while encoding the input, an error variant will be returned.
+    fn encode_input(&self, input: T, writer: &mut dyn io::Write) -> io::Result<usize>;
+}
+
+impl Encoder<Vec<Event>> for (Transformer, crate::codecs::Encoder<Framer>) {
+    fn encode_input(
+        &self,
+        mut events: Vec<Event>,
+        writer: &mut dyn io::Write,
+    ) -> io::Result<usize> {
+        let mut encoder = self.1.clone();
+        let mut bytes_written = 0;
+        let batch_prefix = encoder.batch_prefix();
+        writer.write_all(batch_prefix)?;
+        bytes_written += batch_prefix.len();
+        if let Some(last) = events.pop() {
+            for mut event in events {
+                self.0.transform(&mut event);
+                let mut bytes = BytesMut::new();
+                encoder
+                    .encode(event, &mut bytes)
+                    .map_err(|error| io::Error::new(io::ErrorKind::InvalidData, error))?;
+                writer.write_all(&bytes)?;
+                bytes_written += bytes.len();
+            }
+            let mut event = last;
+            self.0.transform(&mut event);
+            let mut bytes = BytesMut::new();
```

Comment:
These repeated allocations are pretty unfortunate.

We should really think about getting away from the Tokio codec trait -- since I think that's why we're dropping down to writing to a `BytesMut` directly? -- so that we can write directly into `writer`.

That would also let us get back to using `as_tracked_write` to avoid having to manually track the individual writes.

Reply:
Good points. Note that this has only been moved from the deleted `src/sinks/util/encoding/codec.rs` file, but happy to put this into a separate issue

Reply:
Yeah, that's fair.

I think I didn't end up reviewing the original work for the intermediate state of old encoders/new encoders otherwise I would have brought it up there... but I at least don't want to forget about this, so it's more of a "we should follow up on this in the future" comment than anything else.